File: Algos & templates/题目/图论/树与树形图/LC2608.java
class Solution {
    static List<Integer>g;
    static int []dis;
    public static int bfs(int s) {
        int ans = Integer.MAX_VALUE;
        Arraays.fill(dis, -1);
        dis[s] = 0;
        var q = new ArrayDeque<int[]>();//java中没有pair类,强制转换为数组
        q.add(new int[] {s, -1});
        while (! q.isEmpty()) {
            var p = q.poll();
            int x = p[0], fa = p[1];
            for (var y : g[x]) {
                if (dis[y] < 0) {
                    dis[y] = dis[x] + 1;
                    q.add(new int[] {y, x});
                } else if (y != fa) {
                    ans = Math.min(ans, dis[x] + dis[y] + 1);
                }
            }
        }
    }

    public int findShortestCycle(int n, int[][] edges) {
        g = new ArrayList[n + 5];
        //Arrays.setAll(g, e->new ArrayList<>());
        for (int i = 0; i < n; i++) {
            g[i] = new ArrayList<Integer>();
        }
        for (var e : edges) {
            int x = e[0], y = e[1];
            g[x].add(y);
            g[y].add(x);
        }
        dis = new int[n];
        int ans = Integer.MAX_VALUE;
        for (int i = 0; i < n; i++) {
            ans = Math.min(ans, bfs(i));
        }
        return ans < Integer.MAX_VALUE ? ans : -1;
    }
}

File: 图论/分层树 基环树/LC2608.cpp
/*dfs生成树做法求无向图环*/
class Solution {
public:
	int findShortestCycle(int n, vector<vector<int>>& edges) {
		vector<int>tr[n + 5];
		for (auto k : edges) {
			tr[k[0] + 1].push_back(k[1] + 1);
			tr[k[1] + 1].push_back(k[0] + 1);
		}
		vector<bool>vis(n + 5, false);
		vector<int>dep(n + 5, 0), fa(n + 5, 0);
		int ans = 0x3f3f3f3f;
		function<void(int, int)>dfs = [&](int u, int f) {
			vis[u] = 1;
			for (auto v : tr[u]) {
				if (not vis[v]) {
					fa[v] = u;
					dep[v] = dep[u] + 1;
					dfs(v, u);
				} else if (v != f) {
					if (dep[v] < dep[u]) {
						int tot = 1;
						for (int j = u; j != v; j = fa[j]) {
							tot++;
						}
						ans = min(ans, tot);
					}
				}
			}
		};
		for (int i = 1; i <= n; i++) {
			if (not vis[i])dfs(i, 0);
		}
		if (ans == 0x3f3f3f3f) return -1;
		return ans;
	}
};